pkg/controllers/ippool/ippool_controller.go

```diff
@@ -0,0 +1,183 @@
+/* Copyright © 2022 VMware, Inc. All Rights Reserved.
+   SPDX-License-Identifier: Apache-2.0 */
+
+package ippool
+
+import (
+	"context"
+	"runtime"
+	"time"
+
+	apimachineryruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha2"
+	"github.com/vmware-tanzu/nsx-operator/pkg/controllers/common"
+	"github.com/vmware-tanzu/nsx-operator/pkg/logger"
+	"github.com/vmware-tanzu/nsx-operator/pkg/metrics"
+	servicecommon "github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/ippool"
+)
+
+var (
+	log           = logger.Log
+	resultNormal  = common.ResultNormal
+	resultRequeue = common.ResultRequeue
+	MetricResType = common.MetricResTypeIPPool
+)
+
+// Reconciler reconciles a IPPool object
+type Reconciler struct {
+	client.Client
+	Scheme  *apimachineryruntime.Scheme
+	Service *ippool.IPPoolService
+}
+
+func containsString(source []string, target string) bool {
+	for _, item := range source {
+		if item == target {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	obj := &v1alpha2.IPPool{}
+	log.Info("reconciling ippool CR", "ippool", req.NamespacedName)
+	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
+		log.Error(err, "unable to fetch ippool CR", "req", req.NamespacedName)
+		return resultNormal, client.IgnoreNotFound(err)
+	}
+	if obj.ObjectMeta.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(obj, servicecommon.IPPoolFinalizerName) {
+			controllerutil.AddFinalizer(obj, servicecommon.IPPoolFinalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				return resultRequeue, err
+			}
+			log.V(1).Info("added finalizer on ippool CR", "ippool", req.NamespacedName)
+		}
+
+		subnetCidrUpdated, ipPoolSubnetsUpdated, err := r.Service.CreateOrUpdateIPPool(obj)
```

Comment:
I have tried several ways to do this logic and found this one is the most concise, this is different from security policy.
subnetCidrUpdated and ipPoolSubnetsUpdated are very vital to the status updated and Service.FullyRealized(obj) is also vital to control if we would put back obj to queue since some subnets are not realized. In this way, we can return partial realized subnets to user in advance even though partial are not realized. We leverage resultRequeue to implement gradually realizing.

Reply:
I still don't get this. We only need to check if subnets are fully realized, if yes, update status to success, in other cases, requeue to reconcile again.  If ipPoolSubnetsUpdated true but subnetCidrUpdated false, I think we do not need to update IPPool CR again.